KAFKA-2594 Added InMemoryLRUCacheStore

rhauch opened the pull request:
Added a new `KeyValueStore` implementation called `InMemoryLRUCacheStore` that keeps a maximum number of entries in-memory, and as the size exceeds the capacity the least-recently used entry is removed from the store and the backing topic. Also added unit tests for this new store and the existing `InMemoryKeyValueStore` and `RocksDBKeyValueStore` implementations. A new `KeyValueStoreTestDriver` class simplifies all of the other tests, and can be used by other libraries to help test their own custom implementations.

This PR depends upon [KAFKA-2593](https://issues.apache.org/jira/browse/KAFKA-2593) and its PR at https://github.com/apache/kafka/pull/255. Once that PR is merged, I can rebase this PR if desired.

Two issues were uncovered when creating these new unit tests, and both are also addressed as separate (small) commits in this PR:
- The `RocksDBKeyValueStore` initialization was not creating the file system directory if missing.
- `MeteredKeyValueStore` was casting to `ProcessorContextImpl` to access the `RecordCollector`, which prevent using `MeteredKeyValueStore` implementations in tests where something other than `ProcessorContextImpl` was used. The fix was to introduce a `RecordCollector.Supplier` interface to define this `recordCollector()` method, and change `ProcessorContextImpl` and `MockProcessorContext` to both implement this interface. Now, `MeteredKeyValueStore` can cast to the new interface to access the record collector rather than to a single concrete implementation, making it possible to use any and all current stores inside unit tests.

rhauch commented:
Looks like there's an occasional test failure in `core`:

```
542 tests completed, 1 failed
:core:test FAILED

FAILURE: Build failed with an exception.
```

rhauch commented:
@guozhangwang, I've updated this PR to involve a single commit that does nothing but add the `InMemoryLRUCacheStore`. It's based on the updated #255 (which is where the first 4 commits come from), so once that's merged I can rebase this PR.

eribeiro commented on streams/src/main/java/org/apache/kafka/streams/state/InMemoryLRUCacheStore.java:
```diff
@@ -0,0 +1,295 @@
+     * @param time the time provider; may not be null
+     * @return the key-value store
+     */
+    public static <K, V> InMemoryLRUCacheStore<K, V> create(String name, int capacity, ProcessorContext context,
```
@rhauch I am wondering if this class could not benefit from a static `Builder` class. Something along the lines of 

```
InMemoryLRUCacheStore cacheStore = InMemoryLRUCacheStore.Builder.create()
                                                                .name(name)
                                                                .capacity(capacity)
                                                                .context(context)
                                                                .keySerDe(keySerializer, keyDeserializer)
                                                                .valueSerDe(valueSerializer, valueDeserializer)
                                                                .time(time)
                                                                .build();
```

OTOH, the constructor arguments are not the same type (say, String), so no good chances of inverting parameter order or so. Wdyt?

rhauch replied:
Yeah, it'd be nice, but `KeyValueStore` is parameterized, and that makes builders complicated (but not impossible). See the discussion on #255.

guozhangwang commented:
@rhauch Do you think the new org.apache.kafka.common.cache.LRUCache can be reused here?

rhauch commented:
@guozhangwang, the new `org.apache.kafka.common.cache.LRUCache` class just wraps a `LinkedHashMap`, whereas we're using one directly. Also, we require a hook into the cache to respond when an entry is removed, and the `org.apache.kafka.common.cache.LRUCache` doesn't currently provide a way to do that.

rhauch commented:
@guozhangwang, rebased this PR.

guozhangwang commented:
This PR seems still contains some changes from #255, could you double check?

rhauch commented:
@guozhangwang, yes, it is currently based upon #255, so after #255 is merged I will rebase this so it is a single commit.

rhauch committed:
KAFKA-2594 Added InMemoryLRUCacheStore

Added a new KeyValueStore implementation that keeps a maximum number of entries in-memory, and as the size exceeds the capacity the least-recently used entry is removed from the store and the backing topic. Also added unit tests for this new store.

rhauch commented:
@guozhangwang: I've rebased and cleaned up this PR, and it now has only one commit that adds the in-memory cache via the new method in the Stores builder. A local build succeeds with this PR.

Creating a limited-size store is nearly identical to creating an unlimited-sized in-memory store. The difference is that the interface returned by `inMemory()` now has an optional method `maxEntries(int)` with which the caller can specify the maximum number of entries the store should allow:

```
    KeyValueStore<Integer, String> store = Stores.create("my-store", driver.context())
                                                 .withIntegerKeys().withStringValues()
                                                 .inMemory()
                                                 .maxEntries(2047)
                                                 .build();
```

The default maximum number of entries is `Integer.MAX_VALUE`, and that corresponds to an unlimited in-memory key-value store.

guozhangwang commented:
I think this is good since for any in-memory stores, we need some general mechanism to bound the memory usage. LRU is a good starting point to do it, but moving forward we may need to 1) bound the capacity based on bytes instead of number of records (we have seen quite many cases where bounding footprint with number of messages brings ops problems with variable record length), and 2) have customizable evicting methods beyond LRU.

rhauch commented:
@guozhangwang: I completely agree. Perhaps this is a good start, though.

guozhangwang commented:
Yeah just trying to remind ourselves of generalizing it in the future, this patch LGTM.

b-goyal commented:
Added a new `KeyValueStore` implementation called `InMemoryLRUCacheStore` that keeps a maximum number of entries in-memory, and as the size exceeds the capacity the least-recently used entry is removed from the store and the backing topic. Also added unit tests for this new store and the existing `InMemoryKeyValueStore` and `RocksDBKeyValueStore` implementations. A new `KeyValueStoreTestDriver` class simplifies all of the other tests, and can be used by other libraries to help test their own custom implementations.

This PR depends upon [KAFKA-2593](https://issues.apache.org/jira/browse/KAFKA-2593) and its PR at https://github.com/apache/kafka/pull/255. Once that PR is merged, I can rebase this PR if desired.

Two issues were uncovered when creating these new unit tests, and both are also addressed as separate (small) commits in this PR:
- The `RocksDBKeyValueStore` initialization was not creating the file system directory if missing.
- `MeteredKeyValueStore` was casting to `ProcessorContextImpl` to access the `RecordCollector`, which prevent using `MeteredKeyValueStore` implementations in tests where something other than `ProcessorContextImpl` was used. The fix was to introduce a `RecordCollector.Supplier` interface to define this `recordCollector()` method, and change `ProcessorContextImpl` and `MockProcessorContext` to both implement this interface. Now, `MeteredKeyValueStore` can cast to the new interface to access the record collector rather than to a single concrete implementation, making it possible to use any and all current stores inside unit tests.
